GrandCentralBoard/Widgets/WebsiteAnalytics/WebsiteAnalyticsWidget.swift

```diff
@@ -9,44 +9,99 @@ import GCBCore
     }
 
     init(sources: [UpdatingSource]) {
         self.widgetView = TableWidgetView.fromNib()
         self.sources = sources
 
-        let viewModel = WidgetTemplateViewModel(title: "BLOGPOSTS", subtitle: "MOST POPULAR", contentView: widgetView)
+        let viewModel = WidgetTemplateViewModel(title: "BLOGPOSTS".localized.uppercaseString,
+                                                subtitle: "MOST POPULAR".localized.uppercaseString,
+                                                contentView: widgetView)
         let layoutSettings = WidgetTemplateLayoutSettings(contentMargin: UIEdgeInsets(top: -90, left: 0, bottom: 0, right: 0))
-        widgetViewWrapper = WidgetTemplateView.viewWithViewModel(viewModel, layoutSettings: layoutSettings)
+        let widgetViewWithHeader = WidgetTemplateView.viewWithViewModel(viewModel, layoutSettings: layoutSettings)
+
+        mainView = UIView(frame: widgetViewWithHeader.frame)
+        mainView.fillViewWithView(widgetViewWithHeader, animated: false)
     }
 
+
+    // MARK: HavingSources
+
+    let sources: [UpdatingSource]
+
+
+    // MARK: WidgetControlling
+
+    var view: UIView {
+        return mainView
+    }
+
+
+    // MARK: Updateable
+
     func update(source: UpdatingSource) {
         switch source {
-        case let source as PageViewsSource:
-            fetchAnalyticsFromSource(source)
-        default:
-            assertionFailure("Expected `source` as instance of `GoogleAnalyticsSource`.")
+            case let source as PageViewsSource:
+                fetchAnalyticsFromSource(source)
+            default:
+                assertionFailure("Expected `source` as instance of `GoogleAnalyticsSource`.")
         }
     }
 
     private func fetchAnalyticsFromSource(source: PageViewsSource) {
         source.read { [weak self] result in
             switch result {
-            case .Success(let reports):
-                self?.updateWithReports(reports)
-            case .Failure:
-                break
+                case .Success(let reports):
+                    self?.state = reports.isEmpty ? .NoData : .Data(reports: reports)
+                case .Failure:
+                    self?.state = .Error
             }
         }
     }
 
+    private func renderErrorView() {
+        guard !mainView.subviews.contains(errorView) else { return }
+        mainView.fillViewWithView(errorView, animated: false)
+    }
+
+    private func removeErrorView() {
+        errorView.removeFromSuperview()
+    }
+
+
+    // MARK: State rendering
+
+    private func changeFromState(fromState: State, toState: State) {
+        switch toState {
+            case .Error:
+                guard !mainView.subviews.contains(errorView) else { return }
+                mainView.fillViewWithView(errorView, animated: false)
+            case .NoData:
+                guard !mainView.subviews.contains(noDataView) else { return }
+                mainView.fillViewWithView(noDataView, animated: false)
```

Comment:
I think if there was an error, than no data, and then again an error, the view will still be showing 'No data', because it contains'Error' view, but it is hidden under 'No Data'.